### ISSUE TITLE
Notes on intial build I did

### ISSUE THREAD

**Issue by gbowne1** [taotnpwaft/OSDev#6]
Here's what I got on my first build after cloning..

```
$ ./build.sh clang
src/c/stdio.c:7:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
fg_color = BLACK;
^
src/c/stdio.c:8:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
bg_color = BLACK;
^
src/c/stdio.c:33:25: warning: unused parameter 'num' [-Wunused-parameter]
void print_num(uint32_t num) {
                        ^
3 warnings generated.
src/c/stdlib.c:4:18: warning: unused parameter 'size' [-Wunused-parameter]
void* malloc(int size) {
                 ^
1 warning generated.
src/c/string.c:8:12: warning: array subscript is of type 'char' [-Wchar-subscripts]
    if (str[i] != '\0') {
           ^~
1 warning generated.
/usr/bin/ld: cannot open output file bin/kernel.bin: No such file or directory
collect2: error: ld returned 1 exit status
clang: error: linker (via gcc) command failed with exit code 1 (use -v to see invocation)
cp: cannot stat 'bin/kernel.bin': No such file or directory
```

**Comment by gbowne1** [comment]
```
$ ./build.sh gcc
./build.sh: line 36: /home/joey/bin/bin/i686-elf-gcc: No such file or directory
cp: cannot stat 'bin/kernel.bin': No such file or directory
```

**Comment by EmptyGroceryBag** [comment]
> ```
> $ ./build.sh gcc
> ./build.sh: line 36: /home/joey/bin/bin/i686-elf-gcc: No such file or directory
> cp: cannot stat 'bin/kernel.bin': No such file or directory
> ```

Oops. I forgot to remove the absolute path. Ideally the user would have this version of gcc in their `PATH`.

**Comment by EmptyGroceryBag** [comment]
> ```
> /usr/bin/ld: cannot open output file bin/kernel.bin: No such file or directory
> collect2: error: ld returned 1 exit status
> clang: error: linker (via gcc) command failed with exit code 1 (use -v to see invocation)
> cp: cannot stat 'bin/kernel.bin': No such file or directory
> ```
This is because the bin folder doesn't exist after a fresh clone. Maybe I should modify the `.gitignore` file to only ignore `*.bin` files instead of the entire bin directory.

**Comment by gbowne1** [comment]
Yeah.. Makefile or some other thing as well? 

I have both clang and gcc @ latest.

**Comment by gbowne1** [comment]
It did add some .o files though.  

I need to add some .vs & .vscode and .github stuff.

**Comment by EmptyGroceryBag** [comment]
> Yeah.. Makefile or some other thing as well?
> 
> I have both clang and gcc @ latest.
I'm no good at writing Makefiles lol, so you can write one to replace the build script if you want. Also, I don't think regular gcc will work, but you might know about some special flags that I don't.

> It did add some .o files though.
I fixed this bug by removing the `&&` after the linker command in the build script. The `.o` files are deleted whether or not linking was successful.

**Comment by gbowne1** [comment]
Alright. No rush.  

Yeah the build didn't finish.. but it did get far enough to make the .o files.

I am no good at making Makefiles either, but it should be only an option.  Not everyone likes or uses them, but it should be available for those that do.

I will post my gcc and clang versions when I get back to my office.

**Comment by gbowne1** [comment]
Back.  My version of clang:

```
$ clang --version
Debian clang version 11.0.1-2
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

and gcc..

```
$ gcc --version
gcc (Debian 10.2.1-6) 10.2.1 20210110
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by EmptyGroceryBag** [comment]
> Back. My version of clang:
> 
> ```
> $ clang --version
> Debian clang version 11.0.1-2
> Target: x86_64-pc-linux-gnu
> Thread model: posix
> InstalledDir: /usr/bin
> ```

I've only tested clang 14.0.6, but I see no reason why it shouldn't compile on earlier versions.

> ```
> $ gcc --version
> gcc (Debian 10.2.1-6) 10.2.1 20210110
> Copyright (C) 2020 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> ```
> 
> I have g++ too.. which is the same version as my gcc. I believe they are installed together.

Yeah, I don't think this will work. According to this page of OSDev Wiki, only a gcc that targets generic i386 will work properly: https://wiki.osdev.org/GCC_Cross_Compiler

I just pushed some fixes to branch `iss_6`. I cloned a fresh copy of the repo and I was able to build everything from scratch successfully. Let me know if it works with clang 11.

**Comment by gbowne1** [comment]
I think this is just a matter of setting targets for Clang 11.
The only other thing would be to include a CMake and set it up for i386 target.

https://clang.llvm.org/docs/CrossCompilation.html
https://releases.llvm.org/11.0.0/tools/clang/docs/DriverInternals.html
https://stackoverflow.com/questions/28640585/build-install-llvm-clang-with-both-32-64-bit-support-libraries

I also have llvm 11 as well but it does not support any earlier than Pentium Pro. :-1:  at least accoring to llvm --version.

**Comment by gbowne1** [comment]
Deleted /OSDev and Re Cloned. 

It seems to have built.. here is the output I got.
```
./build.sh clang
src/c/kernel.c:6:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
fg_color = WHITE;
^
src/c/kernel.c:7:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
bg_color = BLACK;
^
2 warnings generated.
src/c/stdlib.c:4:18: warning: unused parameter 'size' [-Wunused-parameter]
void* malloc(int size) {
                 ^
1 warning generated.
/usr/bin/ld: boot.o: warning: relocation in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
```

**Comment by gbowne1** [comment]
gcc output
```
$ ./build.sh gcc
./build.sh: line 42: i686-elf-gcc: command not found
```

**Comment by EmptyGroceryBag** [comment]
> gcc output
> 
> ```
> $ ./build.sh gcc
> ./build.sh: line 42: i686-elf-gcc: command not found
> ```

Yes, you need to build a cross compiler for `1686-elf`. In doing so, the Makefile automatically adds the target to the name of the executable.

**Comment by EmptyGroceryBag** [comment]
> I think this is just a matter of setting targets for Clang 11. The only other thing would be to include a CMake and set it up for i386 target.

CMake is indeed an option.

> I also have llvm 11 as well but it does not support any earlier than Pentium Pro. 👎 at least accoring to llvm --version.

So you're saying you'd like to use an earlier clang version to support older Intel CPUs?

**Comment by gbowne1** [comment]
I'm more likely to use clang than gcc for this.  But yeah, you'd have to make a cross compiler.  That process should be in our docs.

I need to update my clang but yes, my notes say that 11 does not support older CPU's.  I could be wrong though.  (latest is Clang 14).  The Linux distro I have uses a repo for packages that is/are way out of date.  even nightly doesn't have recent packages.

This is also why I thought about making a LFS project.

I do have a CMakeLists.txt that could work for a i386 target.. with some work done to it.

I never finished this project.. https://github.com/gbowne1/88os there might be some reuseable sutff in there.

I'm gonna make a post with some thoughts in Discussions.

**Comment by gbowne1** [comment]
@EmptyGroceryBag Its been ages well a whole pile of months and I haven't had a lot of time but I was able to fix a bunch of stuff this weekend 7/7/23 - 7/10/23.  See if you can get it to work with my changes.
